### PR TITLE
[IMP] Field Services Global

### PR DESCRIPTION
**Pull request by iambenkis**
#### 1. Contact form modification
Add a new address type: "Service Location"
Include a star button to display the count of "Service Locations" associated with each company or individual.
Fields for "Service Location" info:
Contact Name
+ phone 
+ email
Equipment name
Address (street, city, country, the regular one)
Type: select (Solar, Water Heater, Genset, Air Conditioning, other) if other is selected, a text field to write i)
Internal Notes

#### 2. Sales Order (SO) updates
Integrate the "Service Location" field below the other address
Add a "Services Instructions" tab:
This tab is visible only if a service product creating a task is added to the SO
field is an HTML one
Make the "Service Instructions" mandatory

#### 3. Task generation and Management
On the task generate upon SO validation, add a tag by default, the name of the Warehouse linked to the Sale Order.
Duplicate the content from "Services Instructions" into the task's description.
Below the instructions, include a table listing materials (product name, product description, and quantity ordered).
Add a "Service Location" field on the task (below other address).
Add a star button linked to the deliveries + display delivery status on task.
Display the task ID as a read-only field .
On the report, display the task ID.

### CODE REVIEW

File: models/res_partner.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+from odoo import fields, models, api
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+    _description = 'Extended Partner Model for Field Service usage'
+    
+    type = fields.Selection(selection_add=[('field_service', 'Service Location')],
+                            help="Service Location: This field represents the type of service.")
+    field_service_type = fields.Selection([
```
Comment (sbuhl) [inline models/res_partner.py:9]: @iambenkis  should be mandatory

File: models/res_partner.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
+from odoo import fields, models, api
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+    _description = 'Extended Partner Model for Field Service usage'
+    
+    type = fields.Selection(selection_add=[('field_service', 'Service Location')],
+                            help="Service Location: This field represents the type of service.")
```
Comment (sbuhl) [inline models/res_partner.py:8]: @iambenkis  "Use to identify local service sites of the customer."

File: models/res_partner.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        ('solar', 'Solar'),
+        ('water_heater', 'Water Heater'),
+        ('genset', 'Genset'),
+        ('air_conditioning', 'Air Conditioning'),
+        ('other', 'Other')],
+        string='Type',
+        help='Select the type of field service'
+    )
+
+    other_description = fields.Char('Other Description')
```
Comment (sbuhl) [inline models/res_partner.py:19]: @iambenkis  "Site Description"

File: models/res_partner.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    
+    @api.onchange('field_service_type')
+    def on_change_field_service_type(self):
+        if self.field_service_type != 'other':
+            self.other_description = ''
```
Comment (sbuhl) [inline models/res_partner.py:24]: @iambenkis  is it necessary? Why not just hide it in XML?

File: models/sale_order.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
+# -*- coding: utf-8 -*-
+
+from odoo import models, fields, api
+
+class SaleOrder(models.Model):
+    _inherit = 'sale.order'
+
+    instruction = fields.Html(string='Service Location')
```
Comment (sbuhl) [inline models/sale_order.py:8]: @iambenkis  String= "Service Instructions"

File: models/sale_order.py [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+from odoo import models, fields, api
+
+class SaleOrder(models.Model):
+    _inherit = 'sale.order'
+
+    instruction = fields.Html(string='Service Location') 
+
+    product_id = fields.Many2one(
+        comodel_name='product.product',related='order_line.product_id',
+        string="Product")
```
Comment (sbuhl) [inline models/sale_order.py:12]: @iambenkis  🤔  something is missing in this model...

File: reports/project_report_views.xml [hunk 7]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<odoo>
+    <template id="worksheet_custom_page_service" inherit_id="industry_fsm_report.worksheet_custom_page">
+        <xpath expr="//h2" position="replace">
+            <h2 t-if="doc.allow_billable and doc.sale_order_id">
+                <span>Time&amp; Material</span>
+                (Task ID: #<span t-field="doc.id"/>) 
+            </h2>
```
Comment (sbuhl) [inline reports/project_report_views.xml:8]: @iambenkis  not the good position. If there is no timesheet, we don't see the Task ID...

File: views/res_partner_views.xml [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+            <field name="model">res.partner</field>
+            <field name="inherit_id" ref="base.view_partner_form"/>
+            <field name="arch" type="xml">
+                    <xpath expr="//notebook//form//sheet" position="replace">
+                        <sheet>
+                            <field name="type" required="1" widget="radio" options="{'horizontal': true}"/>
+                            <field name="parent_id" invisible="1"/>
+                            <div class="text-muted oe_edit_only">
+                                <p class="mb-0" attrs="{'invisible': [('type', '!=', 'contact')]}">
+                                    <span>Use four this to organize the contact details of employees of a given company (e.g. CEO, CFO, ...).</span>
```
Comment (sbuhl) [inline views/res_partner_views.xml:16]: @iambenkis  are we forced to replace the whole file? 
Can't you target an element more precisely to only add what's missing?

**Review by sbuhl — CHANGES_REQUESTED** [review]
@iambenkis  first round of review.

File: models/project.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    name = fields.Text(related='sale_order_line_id.name')
+    order_id = fields.Many2one(related='sale_order_line_id.order_id')
+    order_line = fields.One2many(
+        comodel_name='sale.order.line', related='sale_order_id.order_line')
+    
+    delivery_count = fields.Integer(related='sale_order_id.delivery_count')
+    picking_ids = fields.One2many(related='sale_order_id.picking_ids')
+    partner_id = fields.Many2one(related='sale_order_id.partner_id')
+    
+    tag_ids = fields.Many2many('project.tags', string='Tags', compute='_compute_default_tags', readonly=False)
```
Comment (sbuhl) [inline models/project.py:22]: @iambenkis  look how the value are get from the SO to create the name, customer, etc
https://github.com/odoo/odoo/blob/3652f90e610a786092236ab6e7ca8b3af3d58248/addons/sale_project/models/sale_order_line.py#L215
Maybe some things to use to create the tags...

Comment (iambenkis) [reply models/project.py:22]: Lemme look at it 👀

Comment (sbuhl) [reply models/project.py:21]: Hello @iambenkis,

I've recently conducted a test using Odoo's default method, and I'm pleased to report that the system is effectively handling the tags (add or remove tags). Could you take a closer look and explore how we might adapt your code to ensure seamless operation within Odoo framework?

File: models/res_partner.py (hunk 1 above)
Comment (sbuhl) [reply models/res_partner.py:9]: @iambenkis  do you read my comments?

**Review by sbuhl — CHANGES_REQUESTED** [review]

File: models/res_partner.py (hunk 2 above)
Comment (sbuhl) [reply models/res_partner.py:8]: @iambenkis  ??

File: models/res_partner.py (hunk 1 above)
Comment (sbuhl) [reply models/res_partner.py:9]: @iambenkis  ??

File: models/res_partner.py (hunk 3 above)
Comment (sbuhl) [reply models/res_partner.py:20]: @iambenkis  ??

File: views/res_partner_views.xml [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+                    <xpath expr="//div[@class='o_address_format'][1]" position="after">
+                        <field name='field_service_type' attrs="{'invisible': [('type', '!=', 'field_service')]}" />
+                        <field name="other_description" attrs="{'invisible': [('field_service_type', '!=', 'other')]}"/>
+                        <field name="equipement" attrs="{'invisible': [('type', '!=', 'field_service')]}"/>
+                    </xpath>
+                    
+                    <xpath expr="//notebook//templates//div//div[@class='o_kanban_image']" position="replace">
+                        <div class="o_kanban_image">
+                            <img alt="Contact image" attrs="{'invisible': [('type', '=', 'field_service')]}" t-att-src="kanban_image('res.partner', 'avatar_128', record.id.raw_value)"/>
+                            <img alt="Location image" src="https://drive.google.com/uc?export=view&amp;id=1Gwtc-LN18S89xUfLrUAYtJQpArr-pa--" attrs="{'invisible': [('type', '!=', 'field_service')]}"/>
```
Comment (sbuhl) [inline views/res_partner_views.xml:30]: @iambenkis  pourquoi ne pas utiliser ce qui est dans le folder static?

File: models/sale_order.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
+# -*- coding: utf-8 -*-
+
+from odoo import models, fields, api, _
+from odoo.tools import float_is_zero, format_amount, format_date, html_keep_url, is_html_empty
```
Comment (sbuhl) [inline models/sale_order.py:4]: Those are unused.

File: models/sale_order.py [hunk 12]
```diff
@@ -0,0 +1,40 @@
+# -*- coding: utf-8 -*-
+
+from odoo import models, fields, api, _
```
Comment (sbuhl) [inline models/sale_order.py:3]: is "_" used?

File: models/sale_order.py [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+from odoo.tools import float_is_zero, format_amount, format_date, html_keep_url, is_html_empty
+class SaleOrder(models.Model):
```
Comment (sbuhl) [inline models/sale_order.py:5]: add space before

File: models/sale_order.py [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+            order.has_service = any(line.product_id.type == 'service' and line.product_id.service_tracking == 'task_global_project' for line in order.order_line)
+    
+    @api.depends('partner_id')
+    def _compute_child_id(self):
+        for order in self:
+            n = order.partner_id.child_ids.filtered(lambda child: child.type == 'field_service')
+            if n and len(n) == 1:
+                self.write({'child_id': n.id})
+            elif n and len(n) > 1:
+                self.write({'child_id': ''})
```
Comment (sbuhl) [inline models/sale_order.py:40]: Final new line missing

File: models/res_partner.py [hunk 15]
```diff
@@ -0,0 +1,47 @@
+from odoo import fields, models, api
```
Comment (sbuhl) [inline models/res_partner.py:1]: # -*- coding: utf-8 -*-
Missing, or is it no more needed? Can you check in the doc?

Comment (sbuhl) [reply models/res_partner.py:1]: + "api" is not used.

File: models/res_partner.py [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+            }
+        }
+    
+    def compute_count(self):
+        for record in self:
+            record.service_field_count = self.env['res.partner'].search_count(
+                [
+                    ('type', '=', 'field_service'),
+                    ('id', 'in', self.child_ids.ids)
+                ])
```
Comment (sbuhl) [inline models/res_partner.py:47]: Missing last white line

File: models/project.py [hunk 17]
```diff
@@ -0,0 +1,85 @@
+from odoo import api, fields, models
```
Comment (sbuhl) [inline models/project.py:1]: # -*- coding: utf-8 -*-
 missing

File: models/project.py [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+    tag_ids = fields.Many2many('project.tags', string='Tags', readonly=False)
+    company_id = fields.Many2one(comodel_name='res.company')
+
+    has_default_tag = fields.Boolean(string='Default Tag Created', default=False)
+
+    @api.depends('sale_order_id')
+    def _compute_product_ids(self):
+        for task in self:
+            task.product_ids = task.sale_order_id.order_line.mapped('product_id')
+
```
Comment (sbuhl) [inline models/project.py:29]: remove one line

**Review by sbuhl — CHANGES_REQUESTED** [review]
Hello @iambenkis 
One final review for nitpicking and code convention is needed. Please install a linter to avoid these types of comments in the future.

I'm still waiting for your point of view regarding the addresses.

**Review by sbuhl — APPROVED** [review]

**Pull request merged**